### PR TITLE
#104970: Update and rename perl.xml to perl.md

### LINKED ISSUE #104970: Convert Perl from Docbook to CommonMark
This issue tracks the conversion of `/doc/languages-frameworks/perl.xml` from Docbook to CommonMark per [RFC 72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).

If you intend to work on this conversion, please comment below!

# Instructions for performing this conversion

1. Post in this issue that you intend to work on it. (Please also post if you stop working on it)
2. Checkout a new branch of nixpkgs master.
3. Edit `doc/languages-frameworks/index.xml`: change `perl.xml` to `perl.section.xml`.
4. Move `doc/languages-frameworks/perl.xml` to `doc/languages-frameworks/perl.section.md`.
5. Convert the contents from Docbook to [CommonMark](https://commonmark.org/). Use existing `.md` sections for inspiration.
6. Use `nix-build` from the `doc/` folder to build the manual.
7. Preview the manual on your computer at `http://0.0.0.0:8000/manual.html` after running from `doc/` folder:
  ```
  nix-shell -p python3 --run 'python3 -m http.server --directory ./result/share/doc/nixpkgs'
  ```
8. When it looks good, submit a PR and reference this issue.

### PR DESCRIPTION
**Pull request by StefanSchroeder**
###### Motivation for this change
Addresses #104970, conversion of Docbook language-support for Perl to CommonMark.

@ryantm 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by StefanSchroeder** [comment]
I have deliberately not updated the index.xml, because I expect that the PR will be rejected. But I am interested in the review comments to improve it before actually merging it.

**Comment by ryantm** [comment]
I think you should update the index it makes it easier to test it.

**Comment by ryantm** [comment]
`perl.md` needs to be called `perl.section.md`

File: doc/languages-frameworks/perl.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# Perl
```
Comment (ryantm) [inline doc/languages-frameworks/perl.md:1]:
```suggestion
# Perl {#sec-language-perl}
```
We don't want to break anchor links.

File: doc/languages-frameworks/perl.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+If you are always using the script in places where `nix-shell` is available, you can embed the `nix-shell` invocation in the shebang like this:
+
+```
+#!/usr/bin/env nix-shell
+#! nix-shell -i perl -p perl perlPackages.DBFile
+```
+
+## Packaging Perl programs
+
+   Nixpkgs provides a function *buildPerlPackage*, a generic package builder function for any Perl package that has a standard *Makefile.PL*. It’s implemented in
```
Comment (ryantm) [inline doc/languages-frameworks/perl.md:35]:
```suggestion
   Nixpkgs provides a function `buildPerlPackage`, a generic package builder function for any Perl package that has a standard `Makefile.PL`. It’s implemented in 
```
I think it's more idiomatic in Markdown to use literals instead of emphasis for this kind of thing.

File: doc/languages-frameworks/perl.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+## Packaging Perl programs
+
+   Nixpkgs provides a function *buildPerlPackage*, a generic package builder function for any Perl package that has a standard *Makefile.PL*. It’s implemented in 
+   [pkgs/development/perl-modules/generic](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/perl-modules/generic).
+
+
+   Perl packages from CPAN are defined in
+   [pkgs/top-level/perl-packages.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/perl-packages.nix) rather than `pkgs/all-packages.nix`. Most Perl packages are so straight-forward to build that they are defined here directly, rather than having a separate function for each package called from `perl-packages.nix`. However, more complicated packages should be put in a separate file, typically in `pkgs/development/perl-modules`. Here is an example of the former:
+
+```
```
Comment (ryantm) [inline doc/languages-frameworks/perl.md:42]:
```suggestion
```nix
```

Bunch of places need this change.

**Review by ryantm — CHANGES_REQUESTED** [review]

File: doc/languages-frameworks/perl.md (hunk 2 above)
Comment (jtojnar) [reply doc/languages-frameworks/perl.md:35]: Also the indentation looks weird.

**Comment by StefanSchroeder** [comment]
I have updated the files (index.xml & perl.section.md) but I don't know what to do next.
Do I need to file another PR?

**Comment by ryantm** [comment]
> Do I need to file another PR?

No, your changes got reflected in this one. I'll take a look in a bit, thanks!

File: doc/languages-frameworks/perl.section.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  postInstall = stdenv.lib.optional stdenv.isDarwin ''
+    shortenPerlShebang $out/bin/exiftool
+  '';
+};
+```
+
+This will remove the `-I` flags from the shebang line, rewrite them in the `use lib` form, and put them on the next line instead. 
+This function can be given any number of Perl scripts as arguments; it will modify them in-place.
+
+## Generation from CPAN
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:146]:
```suggestion
## Generation from CPAN {#ssec-generation-from-CPAN}
```

File: doc/languages-frameworks/perl.section.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    meta = {
+      description = "An API for simple XML files";
+      license = with stdenv.lib.licenses; [ artistic1 gpl1Plus ];
+    };
+  };
+```
+
+The output can be pasted into `pkgs/top-level/perl-packages.nix` or wherever else you need it.
+
+## Cross-compiling modules
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:174]:
```suggestion
## Cross-compiling modules {#ssec-perl-cross-compilation}
```

File: doc/languages-frameworks/perl.section.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+```
+
+If you are always using the script in places where `nix-shell` is available, you can embed the `nix-shell` invocation in the shebang like this:
+
+```ShellSession
+#!/usr/bin/env nix-shell
+#! nix-shell -i perl -p perl perlPackages.DBFile
+```
+
+## Packaging Perl programs
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:33]:
```suggestion
## Packaging Perl programs {#ssec-perl-packaging}
```

File: doc/languages-frameworks/perl.section.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
+# Perl {#sec-language-perl}
+
+## Running perl programs on the shell
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:3]:
```suggestion
## Running perl programs on the shell {#ssec-perl-running}
```

**Comment by StefanSchroeder** [comment]
I hadn't seen the "Commit suggestion" button before. That's so much quicker!

File: doc/languages-frameworks/perl.section.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+  };
+
+  preConfigure = ''
+    echo "LIB = ${db.out}/lib" > config.in
+    echo "INCLUDE = ${db.dev}/include" >> config.in
+  '';
+}
+```
+  
+Dependencies on other Perl packages can be specified in the ´buildInputs´ and ´propagatedBuildInputs´ attributes.
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:104]: Somehow backticks got turned to forward ticks.

File: doc/languages-frameworks/perl.section.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+```ShellSession
+$ nix-env -i perl-Class-C3
+```
+
+(Of course you can also install using the attribute name: `nix-env -i -A perlPackages.ClassC3`.)
+
+So what does ´buildPerlPackage´ do? It does the following:
+
+1. In the configure phase, it calls `perl Makefile.PL` to generate a Makefile. You can set the variable `makeMakerFlags` to pass flags to `Makefile.PL`
+2. It adds the contents of the <envar>PERL5LIB</envar> environment variable to `#! .../bin/perl` line of Perl scripts as `-I<replaceable>dir</replaceable>` flags. This ensures that a script can find its dependencies. (This can cause this shebang line to become too long for Darwin to handle; see the note below.)
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:80]: Replaceables have not been replaced.

File: doc/languages-frameworks/perl.section.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  postInstall = stdenv.lib.optional stdenv.isDarwin ''
+    shortenPerlShebang $out/bin/exiftool
+  '';
+};
+```
+
+This will remove the `-I` flags from the shebang line, rewrite them in the `use lib` form, and put them on the next line instead. 
+This function can be given any number of Perl scripts as arguments; it will modify them in-place.
+
+## Generation from CPAN {#ssec-generation-from-CPAN}
```
Comment (jtojnar) [inline doc/languages-frameworks/perl.section.md:146]: This section and the one below previously belonged under “Packaging Perl programs”. Is this change intentional?

**Comment by StefanSchroeder** [comment]
I'll fix it coming weekend.

File: doc/languages-frameworks/perl.section.md (hunk 8 above)
Comment (StefanSchroeder) [reply doc/languages-frameworks/perl.section.md]: That's weird, no idea how that happened. Fixed.

File: doc/languages-frameworks/perl.section.md (hunk 9 above)
Comment (StefanSchroeder) [reply doc/languages-frameworks/perl.section.md]: Fixed <replaceable> and <command>. Multiple instances.

File: doc/languages-frameworks/perl.section.md (hunk 10 above)
Comment (StefanSchroeder) [reply doc/languages-frameworks/perl.section.md]: No, it was not intentional. When looking at the XML source it's difficult to keep track of this kind of detail. It's fixed by increasing the section level from '##' to '###'. Hope that's supported.

**Comment by jtojnar** [comment]
Also, could you please squash the changes?

File: doc/languages-frameworks/perl.section.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+If you are always using the script in places where `nix-shell` is available, you can embed the `nix-shell` invocation in the shebang like this:
+
+```ShellSession
+#!/usr/bin/env nix-shell
+#! nix-shell -i perl -p perl perlPackages.DBFile
+```
+
+## Packaging Perl programs {#ssec-perl-packaging}
+
+Nixpkgs provides a function `buildPerlPackage`, a generic package builder function for any Perl package that has a standard `Makefile.PL`. It’s implemented in
```
[the remark this answers is not in the document]
Comment (StefanSchroeder) [reply doc/languages-frameworks/perl.section.md:35]: I wasn't aware that there is a formatter. I will try it out next.

[the remark this answers is not in the document]
Comment (StefanSchroeder) [reply doc/languages-frameworks/perl.section.md:35]: @jtolnar I am unable to squash my commits. Trying to follow the instructions here https://medium.com/@mittalyashu/how-to-squash-commits-in-a-github-pull-request-97b36576eacb leads to a tremendous amount of merge conflicts by commits that I didn't make... Please advise.
"Auto-merging pkgs/top-level/all-packages.nix
CONFLICT (content): Merge conflict in pkgs/top-level/all-packages.nix
error: could not apply a64d7d84456... Revert "Revert "texinfo: revert to version 6.5 on Darwin""
"

[the remark this answers is not in the document]
Comment (jtojnar) [reply doc/languages-frameworks/perl.section.md:35]: For squashing, I would suggest rebasing onto the parent commit of your first commit instead of `master`. That way you will have less commits to worry about.

```
git rebase -i c7a82285f6c63816b66e78fa1b66fb2bfbe67f8f^
```

But note that you will need to rebase onto `master` anyway, to resolve the conflict in `index.xml`.

**Comment by StefanSchroeder** [comment]
@jtolnar: This is a shoutout to @jtolnar for kindly and patiently guiding me through the commit process of my first contribution. (I don't know what I am doing here. Luckily on the internet no one knows you're a dog.)

**Comment by jtojnar** [comment]
Thanks.

I have rebased it since another merge conflict sneaked in and will merge this after the CI finishes,
